storage/remote/queue_manager.go

```diff
@@ -763,6 +776,58 @@ outer:
 	return true
 }
 
+func (t *QueueManager) AppendWALMetadata(ms []record.RefMetadata) bool {
+	if !t.sendMetadata {
+		return true
+	}
+
+outer:
+	for _, m := range ms {
+		t.seriesMtx.Lock()
+		lbls, ok := t.seriesLabels[m.Ref]
+		if !ok {
+			t.metrics.droppedMetadataTotal.Inc()
+			// Track dropped metadata in the same EWMA for sharding calc.
+			t.dataDropped.incr(1)
+			if _, ok := t.droppedSeries[m.Ref]; !ok {
+				level.Info(t.logger).Log("msg", "Dropped metadata for series that was not explicitly dropped via relabelling", "ref", m.Ref)
+			}
+			t.seriesMtx.Unlock()
+			continue
+		}
+		t.seriesMtx.Unlock()
+
+		backoff := t.cfg.MinBackoff
+		for {
+			select {
+			case <-t.quit:
+				return false
+			default:
+			}
+			if t.shards.enqueue(m.Ref, timeSeries{
+				seriesLabels: lbls,
+				sType:        tMetadata,
+				metadata: &metadata.Metadata{
+					Type: record.ToTextparseMetricType(m.Type),
+					Help: m.Help,
+					Unit: m.Unit,
+				},
+			}) {
+				continue outer
+			}
+
+			t.metrics.enqueueRetriesTotal.Inc()
+			time.Sleep(time.Duration(backoff))
+			backoff *= 2
+			if backoff > t.cfg.MaxBackoff {
+				backoff = t.cfg.MaxBackoff
+			}
+		}
+	}
+
+	return true
+}
```

Comment:
I think actually this should change. If we have metadata records in the WAL and we're going to be sending metadata alongside series labels and samples in the same request, we should be treating metadata records the same as we do series records. That is, we should cache the metadata for a series, check for it whenever we receive a sample that we want to put into the next batch, and also clean up that cache when we clean up the series label cache.

Reply:
So if I understand correctly, instead of enqueueing a new timeSeries struct every time a WAL metadata record is updated, we'd like to send metadata along with _every_ sample we receive for a series?

Sending them continuously is closer to what we do now (collect and send everything every 1m by default), and the truth is that we'll never send a metadata record _without_ a sample since they only appear in the scrape loop.

I'm just not sure whether that's wasteful or not.

I've went ahead to implement this on bd41c24e4237838f30d1e7e701201215f578fa95 if you want to take a look. Is that what you had in mind?
a) Make the `TimeSeries` proto definition contain a single metadata instance (remove the `repeated` keyword)
b) I've removed the `tMetadata seriesType` and the ability to enqueue metadata on its own
c) Use the StoreMetadata method to keep track of metadata similar to the seriesLabels cache and pass that along wherever we create a `timeSeries` struct (in Append/AppendExemplars/AppendHistograms etc).
d) If `sendMetadata` is true, we append this cache value to the outgoing timeSeries struct in `populateTimeSeries`
e) When the series is gone, we clear the cache entry along the series' labels

Reply:
After an offline sync, I'm reverting this approach (at least for now), to avoid making all requests bigger by default.

I'm keeping the change of each TimeSeries struct containing a single Metadata (not repeated).

Reply:
You can't just change the remote-write proto definition, it's a standard implemented by many parties.

Reply:
My initial approach was piggy-backing off the current way metadata is sent over remote_write, but was pretty inefficient and didn't take great advantage of our previous work.

This approach adds a new field, which should be backwards compatible and also similar to how Histograms were also recently added in https://github.com/prometheus/prometheus/commit/d9d51c565c622cdc7d626d3e7569652bc28abe15. Also, like histograms, this is also something opt-in. With these in mind, are there still concerns about changing the proto definition? If then, what would be your suggestion as the next steps, to start a wider discussion?

Reply:
Broadly, write a Proposal and get it accepted. See for example https://github.com/prometheus/prometheus/issues/10539
Metadata is an existing feature, unlike Native Histograms, so the impact of changing the way they work is much greater.

Reply:
As far as I had understood how the change was being proposed was that current "metadata sync" flow would stay the same but that inline metadata would be added (so that it can be identified the exact type of a metric at push time for metric pipeline processing, rather than at sync time which requires a foreign join on metric type which not be 100% accurate at push time per series if the type of the metric changes for instance).

@tpaschalis is that how you had been approaching this or had you also been intending perhaps to remove the current metadata sync exchange (which is sent at 1min interval by default)?

I think @bboreham what @tpaschalis is pointing out that sending the metadata inline (which the whole point of this is as I understand is to make certain attributes available at push time and guaranteed to be available) has far lower network transfer than sending it in the "metadata" field that exists in protobuf today (since then labels for each series are written once, rather than twice, once for timeseries message and once for metadata message). 

As far as I understand though, instead of changing any current metadata sync requests that are made or removing the metadata field in protobuf, it could be simply an additive change to also make metadata available inline on the timeseries message (and hence reduce the overall network effect made by this change).

Here's the original discussion on this approach:
https://github.com/prometheus/prometheus/pull/7771

Related design doc:
https://docs.google.com/document/d/1LY8Im8UyIBn8e3LJ2jB-MoajXkfAqW2eKzY735aYxqo/edit#heading=h.bik9uwphqy3g

Related mailing list discussion:
https://groups.google.com/g/prometheus-developers/c/w-eotGenBGg/m/UZWdxzcBBwAJ

Related stats (more on that PR link):
> Here's the increase with a default config, the difference is relatively less measurable - likely to do with the previous config sending larger batches since the send deadline was longer with a max samples per send of 5,000:
> - steady state ~177kb/sec without this change
> - steady state ~210kb/sec with this change
> - roughly 20% increase
![90612348-e9803a80-e1d5-11ea-95d5-84608012a5a9](https://user-images.githubusercontent.com/339615/235618778-b33d9e11-8470-47ec-8373-b62f66fd641c.png)

Reply:
I apologise, I was reacting to bd41c24e4237838f30d1e7e701201215f578fa95, which includes a change to a field:
https://github.com/prometheus/prometheus/commit/bd41c24e4237838f30d1e7e701201215f578fa95#diff-bdaf80ebc5fa26365f45db53435b960ce623ea6f86747fb8870ad1abc355f64fR150

Now I see that field was only added in the context of this PR; it wouldn't itself be a breaking change.

I'm sorry I haven't read all the history but I don't understand why we are discussing metadata on a per-series level.
Prometheus doesn't ingest or store metadata that way; it's per-name or "metric family".

Reply:
Elsewhere, people have pointed me at a conversation where metadata on a per-series level was discussed.
Please ignore me, I don't have the context to get involved.